src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -55,6 +55,18 @@
 // Disabling pending crashes
 #define ENABLE_CONNECTIVITY_MONITORING 0
 
+#ifdef DEBUG
+#define MTR_REQUIRE_SUBCLASS_IMPL_PASTED(message) \
+    MTR_LOG_ERROR(message);                       \
+    NSAssert(NO, @message)
+#else // DEBUG
+#debug MTR_REQUIRE_SUBCLASS_IMPL_PASTED(message) \
+    MTR_LOG_ERROR(message)
+#endif // DEBUG
+
+#define MTR_REQUIRE_SUBCLASS_IMPL(selectorString) \
+    MTR_REQUIRE_SUBCLASS_IMPL_PASTED("MTRDevice " selectorString " must be handled by subclasses")
+
```

Comment:
```suggestion
#ifdef DEBUG
#define MTR_REQUIRE_SUBCLASS_IMPL_PASTED(message, ...) \
    MTR_LOG_ERROR(message, __VA_ARGS__);                            \
    NSAssert(NO, @message, __VA_ARGS__);
#else // DEBUG
#define MTR_REQUIRE_SUBCLASS_IMPL_PASTED(message, ...) \
    MTR_LOG_ERROR(message, __VA_ARGS__);
#endif // DEBUG

#define MTR_REQUIRE_SUBCLASS_IMPL() \
    MTR_REQUIRE_SUBCLASS_IMPL_PASTED("MTRDevice %s must be handled by subclasses", [NSStringFromSelector(_cmd) cStringUsingEncoding:NSUTF8StringEncoding])

```

Reply:
```suggestion
#define _MTR_REQUIRE_SUBCLASS_IMPL_LOG() \
    MTR_LOG_ERROR("%@ %@ must be implemented by a sub-class", [self class], NSStringFromSelector(_cmd))

#define _MTR_REQUIRE_SUBCLASS_IMPL_THROW() \
    @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"MTR_REQUIRE_SUBCLASS_IMPL" userInfo:nil]

#ifdef DEBUG
#define MTR_REQUIRE_SUBCLASS_IMPL()                 \
    do { _MTR_REQUIRE_SUBCLASS_IMPL_LOG();          \
    _MTR_REQUIRE_SUBCLASS_IMPL_THROW; } while (0)
#else
#define MTR_REQUIRE_SUBCLASS_IMPL() _MTR_REQUIRE_SUBCLASS_IMPL_LOG()
#endif // DEBUG
```

Reply:
Also we can just put this into `MTRDefines_Internal.h`